.ci/ci-get-build-order.py

```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/env python3
+
+from dataclasses import dataclass
+from multiprocessing.pool import ThreadPool
+
+import os
+import subprocess
+import re
+
+
+@dataclass
+class PackageInfo:
+    deps: list
+    processed: bool
+
+
+def run(*args, **kwargs):
+    return subprocess.check_output(args, **kwargs).decode("utf-8").strip()
+
+
+def get_toplevel():
+    path = run("git", "rev-parse", "--show-toplevel")
+    return run("cygpath", "-m", path)
+
+
+def list_changes(*git_args):
+    out = run("git", "log", *git_args, "upstream/master..").splitlines()
+    out += run("git", "log", *git_args, "HEAD^..").splitlines()
+    return list(dict.fromkeys(x.split("::")[-1] for x in sorted(out)))
+
+
+def list_commits():
+    return list_changes("--pretty=format:%ai::[%h] %s")
+
+
+def list_packages():
+    changes = list_changes("--pretty=format:", "--name-only")
+    return [
+        x.split("/")[0]
+        for x in changes
+        if x.endswith("/PKGBUILD") and os.path.exists(x)
+    ]
+
+
+def get_pkginfo(package, packageset):
+    props = ["depends", "makedepends", "pkgname", "provides"]
+    script = f'set -e && source "{package}/PKGBUILD"\n'
+    for prop in props:
+        script += f"echo \"${{{prop}[@]}}\" && printf '\\0'\n"
+
+    shell = os.environ.get("SHELL", "bash")
+    env = {"MINGW_PACKAGE_PREFIX": "mingw-w64"}
+    results = run(shell, "-c", script, env=env).split("\0")[:-1]
+    assert len(props) == len(results), "Length of props matches results"
+
+    info = {}
+    version_re = re.compile(r"[<>=]")
+    for prop, res in zip(props, results):
+        res = [x for x in res.strip().split() if x]
+        if prop in ("depends", "makedepends"):
+            res = [version_re.split(x)[0] for x in res]
+        info[prop] = res
```

Comment:
I've just spent some time parsing pacman databases, which use the same version format...

```suggestion
    version_re = re.compile(r"[<>=]")
    for prop, res in zip(props, results):
        res = [x for x in res.strip().split() if x]
        if prop in ("depends", "makedepends", "provides"):
            res = [version_re.split(x)[0] for x in res]
        info[prop] = res
```

Provides can also specify a version.  Theoretically, they all could also end in ": description", but I've not seen anything but optdepends use that.  I found attempting to split that off using a regex is slower than using `.rsplit(": ", 1)`, if you did want to be robust against that.

Reply:
> Provides can also specify a version

What does this format look like?

Reply:
Ah, I see an example on https://wiki.archlinux.org/title/PKGBUILD#provides

Reply:
yeah.  They're all (provides, depends, optdepends, ...) parsed through the same function in libalpm (library behind pacman), effectively rsplit(': ', 1) to get optional description, followed by taking the first part from that split and splitting on the first of '(<=|>=|<|>|=)', first of that is the name and second is the version.

Reply:
I ended up just passing everything through the regex `[<>=:]` to strip out both in one step.

Reply:
That'll work as long as you only care about the name, not in splitting the version from the description (which you don't, so :+1: